runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -208,71 +212,69 @@ public void setup() throws Exception {
 
     @Override
     public void processElement(final ProcessContext c) {
-      final WindowingInternals<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT> internals =
-          c.windowingInternals();
-      StateInternals<?> stateInternals = internals.stateInternals();
-      TimerInternals timerInternals = internals.timerInternals();
-
+      // Initialize state (element and restriction) depending on whether this is the seed call.
+      // The seed call is the first call for this element, which actually has the element.
+      // Subsequent calls are timer firings and the element has to be retrieved from the state.
+      TimerInternals.TimerData timer = Iterables.getOnlyElement(c.element().timersIterable(), null);
+      boolean isSeedCall = (timer == null);
+      StateNamespace stateNamespace =
+          isSeedCall
+              ? StateNamespaces.window((Coder) windowCoder, c.window())
+              : timer.getNamespace();
       ValueState<WindowedValue<InputT>> elementState =
-          stateInternals.state(StateNamespaces.global(), elementTag);
+          c.windowingInternals().stateInternals().state(stateNamespace, elementTag);
       ValueState<RestrictionT> restrictionState =
-          stateInternals.state(StateNamespaces.global(), restrictionTag);
+          c.windowingInternals().stateInternals().state(stateNamespace, restrictionTag);
       WatermarkHoldState<GlobalWindow> holdState =
-          stateInternals.state(StateNamespaces.global(), watermarkHoldTag);
-
-      elementState.readLater();
-      restrictionState.readLater();
+          c.windowingInternals().stateInternals().state(stateNamespace, watermarkHoldTag);
 
-      final WindowedValue<InputT> element;
-      RestrictionT restriction = restrictionState.read();
-      final boolean wasFirstCall = (restriction == null);
-      if (wasFirstCall) {
-        // This is the first ProcessElement call for this element/restriction.
+      ElementRestriction<WindowedValue<InputT>, RestrictionT> state;
```

Comment:
this seems like it should be called `restriction`, given the accesses you're making to `StateInternals` (`elementState`, `restrictionState`...)